setup/setup_snyk.sh

```diff
@@ -49,48 +53,64 @@ esac
 } > snyk
 
 if ! command -v "$SUDO_CMD" &> /dev/null; then
-  echo "$SUDO_CMD is NOT installed. Trying without sudo, expecting privileges to write to '/usr/local/bin'."
+  echo_with_timestamp "$SUDO_CMD is NOT installed. Trying without sudo, expecting privileges to write to '/usr/local/bin'."
   SUDO_CMD=""
 else
-    echo "$SUDO_CMD is installed."
+    echo_with_timestamp "$SUDO_CMD is installed."
 fi
 
 chmod +x snyk
 ${SUDO_CMD} mv snyk /usr/local/bin
 # Function to download a file with fallback to backup URL
 # Parameters:
-#   $1: File name to download
+#   $1: Download URL
 #   $2: Output file name
 download_file() {
-    # Try to download from the main URL
-    if curl --compressed --retry 2 --output "$2" "$MAIN_URL/$1?utm_source="$GH_ACTIONS; then
-        echo "Downloaded from $MAIN_URL/$1"
-    # If main URL fails, try the backup URL
-    elif curl --compressed --retry 2 --output "$2" "$BACKUP_URL/$1?utm_source="$GH_ACTIONS; then
-        echo "Downloaded from $BACKUP_URL/$1"
-    # If both URLs fail, return an error
+    echo_with_timestamp "Downloading Snyk binary"
+    if curl --fail -D - --compressed --retry 2 --output "$2" "$1/$2?utm_source="$GH_ACTIONS; then
+        echo_with_timestamp "Downloaded from $1/$2?utm_source=$GH_ACTIONS"
+    else
+        echo_with_timestamp "Failed to download from $1/$2?utm_source=$GH_ACTIONS"
+        return 1
+    fi
+
+    echo_with_timestamp "Downloading Snyk binary shasum"
+    if curl --fail -D - --compressed --retry 2 --output "$2.sha256" "$1/$2.sha256?utm_source="$GH_ACTIONS; then
+        echo_with_timestamp "Downloaded from $1/$2.sha256?utm_source=$GH_ACTIONS"
     else
-        echo "Failed to download $1 from both URLs"
+        echo_with_timestamp "Failed to download from $1/$2.sha256?utm_source=$GH_ACTIONS"
         return 1
     fi
 }
 
-# Download Snyk binary
-if ! download_file "$VERSION/snyk-${PREFIX}" "snyk-${PREFIX}"; then
-    die "Failed to download Snyk binary"
+if ! download_file "$MAIN_URL/$VERSION" "snyk-${PREFIX}"; then
+    echo_with_timestamp "Failed to download Snyk files"
+    
+    echo_with_timestamp "Retrying download with secondary URL"
+    if ! download_file "$BACKUP_URL/$VERSION" "snyk-${PREFIX}"; then
+        die "Failed to download Snyk files"
+    fi
 fi
 
-# Download SHA256 checksum file
-if ! download_file "$VERSION/snyk-${PREFIX}.sha256" "snyk-${PREFIX}.sha256"; then
-    die "Failed to download SHA256 file"
+# Verify the checksum
+echo_with_timestamp "Validating shasum"
```

Comment:
Suggestion: After some thought, I do have one suggestion. Could we move the shasum validation into download_file()? This way, we would attempt the fallback if the shasum fails. Also the operation download and shasum check seem reasonable to do together.